src/Storages/transformQueryForExternalDatabaseAnalyzer.cpp

```diff
@@ -0,0 +1,81 @@
+#include <Storages/transformQueryForExternalDatabaseAnalyzer.h>
+
+#include <Storages/transformQueryForExternalDatabase.h>
+#include <Parsers/ASTSelectWithUnionQuery.h>
+#include <Analyzer/InDepthQueryTreeVisitor.h>
+#include <Interpreters/convertFieldToType.h>
+
+#include <Columns/ColumnConst.h>
+
+#include <Analyzer/QueryNode.h>
+#include <Analyzer/ColumnNode.h>
+#include <Analyzer/ConstantNode.h>
+#include <Analyzer/ConstantValue.h>
+#include <Analyzer/TableNode.h>
+#include <Analyzer/JoinNode.h>
+
+
+#include <DataTypes/DataTypesNumber.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int UNSUPPORTED_METHOD;
+}
+
+namespace
+{
+
+class PrepareForExternalDatabaseVisitor : public InDepthQueryTreeVisitor<PrepareForExternalDatabaseVisitor>
+{
+public:
+    using Base = InDepthQueryTreeVisitor<PrepareForExternalDatabaseVisitor>;
+    using Base::Base;
+
+    static void visitImpl(QueryTreeNodePtr & node)
+    {
+        auto * constant_node = node->as<ConstantNode>();
+        if (constant_node)
+        {
+            auto result_type = constant_node->getResultType();
+            if (isDate(result_type) || isDateTime(result_type) || isDateTime64(result_type))
+            {
+                /// Use string representation of constant date and time values
```

Comment:
Via

https://github.com/ClickHouse/ClickHouse/blob/48bba254ad63e65927247b8d3c1821b377d45a2d/src/Storages/transformQueryForExternalDatabase.cpp#L31-L33

It is applied on top of AST generated from query tree, but doesn't cover all cases because in that AST we will have dates as numbers that are not replaced with this visitor.